### PR TITLE
Fix limit logic for EventsStream

### PR DESCRIPTION
**Pull request by richvdh**
As with #7337, this is partly about fixing an assertion which was added recently, but I'm also fixing a much longer-standing bug in the replication catchup logic.

Also adds some tests. The first couple of commits are just some rejigging of the test framework: it might be helpful to review them separately.

### CODE REVIEW

File: tests/replication/tcp/streams/_base.py [hunk 1]
```diff
@@ -37,7 +39,6 @@ def prepare(self, reactor, clock, hs):
         self.server = server_factory.buildProtocol(None)
 
         repl_handler = ReplicationCommandHandler(hs)
-        repl_handler.handler = self.test_handler
```
Comment (richvdh) [inline tests/replication/tcp/streams/_base.py:40]: this was redundant (and caused an error in mypy)

**Comment by richvdh** [comment]
boo, this is going to conflict with #7278.

**Comment by richvdh** [comment]
(conflicts now fixed)

File: synapse/replication/tcp/streams/events.py [hunk 2]
```diff
@@ -176,16 +176,30 @@ async def _update_function(
...
+        # the limit there, we'll need to truncate the results so that we have a complete
+        # set of changes for all the stream IDs we include.
         if len(state_rows) == target_row_count:
             assert state_rows[-1][0] <= upper_limit
-            upper_limit = state_rows[-1][0]
-            limited = True
+            upper_limit = state_rows[-1][0] - 1
+
+            # search for the point to truncate the list
+            for idx in range(len(state_rows) - 1, 0, -1):
```
Comment (erikjohnston) [inline synapse/replication/tcp/streams/events.py:189]: I'd probably write this as `for idx, row in reversed(enumerate(state_rows)):` or something

File: synapse/replication/tcp/streams/events.py [hunk 3]
```diff
@@ -176,16 +176,30 @@ async def _update_function(
...
+                    state_rows = state_rows[:idx]
+                    break
+            else:
+                # bother. We didn't get a full set of changes for even a single
+                # stream id. let's run the query again, without a row limit, but for
+                # just one stream id.
+                upper_limit += 1
+                state_rows = await self._store.get_all_updated_current_state_deltas(
+                    from_token, upper_limit, limit=None
+                )
```
Comment (erikjohnston) [inline synapse/replication/tcp/streams/events.py:200]: I think we should probably move this logic into `get_all_updated_current_state_deltas`, as I think its basically broken if it doesn't return all rows for a given stream ID

File: synapse/replication/tcp/streams/events.py (hunk 2 above)
Comment (richvdh) [reply synapse/replication/tcp/streams/events.py:189]: you can't do `reversed(enumerate(...))`, because enumerate returns a generator.

Comment (erikjohnston) [reply synapse/replication/tcp/streams/events.py:189]: Well that's tedious isn't it. Might be worth saying that we're walking backwards to find the last row of the set of rows with the second to last stream ID. Or something. But with better words. Mebs. :man_shrugging:

**Pull request merged**